NIFI-2028: Fixed Site-to-Site Transit URI

ijokarumawak opened the pull request:
Please review and merge if it looks ok.

I've added unit tests and also did IT test to check https and proxy works without problem.
Transit URIs are more consistent now. Please also refer [NIFI-2028](https://issues.apache.org/jira/browse/NIFI-2028) for detail on what was wrong and how it's fixed.

Thanks!

JPercivall commented:
Hey @ijokarumawak, looks like this has merge conflicts. Could you address them?

ijokarumawak commented:
@JPercivall Fixed conflicts and tested with the latest code base. Please take a look. Thanks!

ijokarumawak commented:
@JPercivall This PR had been conflicted again. Rebased, tested with the latest codebase, and confirmed it works. I hope this can be merged soon.

markap14 commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-site-to-site/src/main/java/org/apache/nifi/remote/StandardRemoteGroupPort.java:
```diff
@@ -366,6 +371,17 @@ private int receiveFlowFiles(final Transaction transaction, final ProcessContext
 
+    private String createTransitUri(final Transaction transaction, String sourceFlowFileIdentifier) {
+        // Use actual uri as transitUri if this transport protocol has meaningful uri.
+        switch (remoteGroup.getTransportProtocol()) {
```
This feels wrong to me. We should be using the URI that the communicant gives to us, regardless of which implementation is being used. If the URI returned by the HTTP implementation is not meaningful, then we should update the code so that we pass in a more meaningful URI

ijokarumawak replied:
Hmm. Theoretically, I agree with that. However, there're couple of reasons I ended up writing code like that:

1. Existing Socket site-to-site transit URI has `sourceFlowFileIdentifier` in it, which can't be a part of the URI that the communicant knows. I think this can be removed since a provenance event has `sourceSystemFlowFileId`. If it's ok to change existing rule, then we can use communicant url as transit url here.

2. There're two URLs in this context, one is Communicant (Peer) url, another is CommunicationsSession's url. For RAW socket, both look the same as "nifi://{hostname}:{port}". For HTTP, Peer.url="http://{hostname}:{port}/nifi-api", CommunicationsSession.url="http://{hostname}:{port}/nifi-api/data-transfer/input-ports/{uuid}/transactions/{tx-id}/flow-files". I thought CommunicationsSession's url is more appropriate for transit url.

Having said that, I think those distinction shouldn't exist here in StandardRemoteGroupPort. I will move those to more protocol specific classes. Thanks @markap14 for your comment!

ijokarumawak commented:
Hi @markap14 Updated the PR.

- For HTTP Site-to-Site, simply use Communicant url as transit uri.
- For RAW Site-to-Site, added CustomTransitUriTransaction to keep using existing format.

Removed TRANSIT_URI HTTP header since transit url can be retrieved from HTTP request object at server side. Also, removed url from CommunicationsSession.

I hope this addresses your feedback. Please take a look at it when you have time. Thank you!

markap14 commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-site-to-site/src/main/java/org/apache/nifi/remote/StandardRemoteGroupPort.java:
```diff
@@ -366,6 +371,14 @@ private int receiveFlowFiles(final Transaction transaction, final ProcessContext
     }
 
+    private String createTransitUri(Transaction transaction, String sourceFlowFileIdentifier) {
+        if (transaction instanceof CustomTransitUriTransaction) {
```
@ijokarumawak I think I like this a little better, but it still feels off to me. Transaction is an interface, so I don't like inspecting the type of the interface in order to determine how to interact with it. What if instead of taking this approach, we update the Communicant interface to have a `createTransitUri(String sourceFlowFileIdentifier)` method? Then, we can just call `transaction.getCommunicant().createTransitUri(sourceFlowFileIdentifier)`. The raw socket transaction and the HTTP Transaction can then wrap the `Peer` class as necessary to ensure that the appropriate URI is created. This allows us to get away from having to have special interfaces that extend Transaction as well as having to inspect the Transaction to find out how to handle it. It also avoids modifying the `Transaction` interface, which is sort of the 'main' interface for Site-to-Site communications. Thoughts?

ijokarumawak replied:
@markap14 Thanks for the comment, I agree with that. I felt that I should have kept Peer completely immutable, so I might bypass the createTransitUri call to CommunicationsSession. Will update the PR soon.

ijokarumawak commented:
@markap14 Thanks again for reviewing! I've updated the PR to move the extension point from Transaction to Communicant. Would you take a look?

JPercivall commented:
Sorry, @ijokarumawak I accidentally mistyped the "close" statement on a PR I merged. Please re-open this.

ijokarumawak commented:
Where can I found the reopen button? Let me test if I can find one by adding a comment.

ijokarumawak commented:
@JPercivall I couldn't find a reopen button on this PR page, so I recreated a PR #751 .
